### DISCUSSION TITLE
Neopixel WS2812 and gpiozero (remote GPIO)

### DISCUSSION THREAD

**Discussion by AL1-1956** [gpiozero/gpiozero#1176] [Q&A]
Hello Everybody

Is there a way to control a WS2812 Led Strip (144 LEDs/meter) connected to a Raspberyy PI 
from a Windows PC with gpiozero and remote GPIO ?

Thanks for reading

Kindly

Alain

**Comment by lurch** [comment]
The WS2812 LED strips need very precise timing, and I suspect that using a simple remote-GPIO approach would introduce too much latency for the WS2812 LEDs to respond correctly?
So it's possible, but you'd probably need to have an application running on the Pi directly controlling the WS2812 LEDs, and some other mechanism (i.e. not remote GPIO) for communicating with this Pi application from the Windows PC.

**Comment by AL1-1956** [comment]
Okay , thank you Andrew for the quick answer (I am French so excuse my poor English)

I have here an extract of a program that runs a colorwipe test of the LED Strip

#!/usr/bin/env python3
# NeoPixel library strandtest example
# Author: Tony DiCola (tony@tonydicola.com)
#
# Direct port of the Arduino NeoPixel library strandtest example.  Showcases
# various animations on a strip of NeoPixels.

import time
from neopixel import *
import argparse

# LED strip configuration:
LED_COUNT      = 16      # Number of LED pixels.
LED_PIN        = 18      # GPIO pin connected to the pixels (18 uses PWM!).
#LED_PIN        = 10      # GPIO pin connected to the pixels (10 uses SPI /dev/spidev0.0).
LED_FREQ_HZ    = 800000  # LED signal frequency in hertz (usually 800khz)
LED_DMA        = 10      # DMA channel to use for generating signal (try 10)
LED_BRIGHTNESS = 100     # Set to 0 for darkest and 255 for brightest
LED_INVERT     = False   # True to invert the signal (when using NPN transistor level shift)
LED_CHANNEL    = 0       # set to '1' for GPIOs 13, 19, 41, 45 or 53



# Define functions which animate LEDs in various ways.
def colorWipe(strip, color, wait_ms=50):
    """Wipe color across display a pixel at a time."""
    for i in range(strip.numPixels()):
        strip.setPixelColor(i, color)
        strip.show()
        time.sleep(wait_ms/1000.0)

# Main program logic follows:
if __name__ == '__main__':
    # Process arguments
    parser = argparse.ArgumentParser()
    parser.add_argument('-c', '--clear', action='store_true', help='clear the display on exit')
    args = parser.parse_args()

    # Create NeoPixel object with appropriate configuration.
    strip = Adafruit_NeoPixel(LED_COUNT, LED_PIN, LED_FREQ_HZ, LED_DMA, LED_INVERT, LED_BRIGHTNESS, LED_CHANNEL)
    # Intialize the library (must be called once before other functions).
    strip.begin()

    print ('Press Ctrl-C to quit.')
    if not args.clear:
        print('Use "-c" argument to clear LEDs on exit')

    try:

        while True:
            print ('Color wipe animations.')
            colorWipe(strip, Color(255, 0, 0))  # Red wipe
            colorWipe(strip, Color(0, 255, 0))  # Blue wipe
            colorWipe(strip, Color(0, 0, 255))  # Green wipe

    except KeyboardInterrupt:
        if args.clear:
            colorWipe(strip, Color(0,0,0), 10)

I think understanding that the LED Strip is controlled by GPIO BCM pin 18 with PWM

So in theory this PWM signal could be sent remotely by the PC to the Pi

Am I right ??

Thank you

Alain

**Reply by lurch** [reply]
> So in theory this PWM signal could be sent remotely by the PC to the Pi
> 
> Am I right ??

I've got no idea, you might want to ask Tony DiCola ? :slightly_smiling_face:  But note that one of the comments says "(usually 800khz)" and I doubt such a high-speed signal would work reliably over a network.

I suspect that you might get better suggestions from https://forums.raspberrypi.com/ as this doesn't seem to be something directly related to GpioZero?

**Reply by dsssssssss9** [reply]
https://cascadememorial.com/obituary/659469/Anthony-Charles-Dicola/

**Reply by lurch** [reply]
:cry:  Sorry, I didn't mean to come across as insensitive.

**Reply by AL1-1956** [reply]
I did not know either about what happened to Tony DiCola.... and I am very sorry for his loved ones...

But maybe I asked the wrong question.

I was thinking something like that.

From the PC give the order , by remote GPIO , to the Pi to send a PWM signal to the LED Strip
so the 800 Khz PWM signal does not run through the network , only the order to send it ...

Maybe I make myself illusions but it would be cool to have a "real" editor like "Thonny" on the PC ,instead of "Nano" on the Pi without 
having to plug a keyboard mouse and monitor to the Pi (to run Thonny on the Pi)

Anyway Thank You 

Kindly

Alain

**Reply by lurch** [reply]
> From the PC give the order , by remote GPIO , to the Pi to send a PWM signal to the LED Strip
> so the 800 Khz PWM signal does not run through the network , only the order to send it ...

Yes, that's entirely possible, but "remote GPIO" (which is just about turning GPIO pins on and off, not controlling high-speed PWM) isn't the mechanism for doing that, which is why I think that GpioZero isn't the answer to your problem.

Oh, although I've just checked and [PiGpio](https://gpiozero.readthedocs.io/en/latest/api_pins.html#module-gpiozero.pins.pigpio) apparently supports both [remote pins](https://gpiozero.readthedocs.io/en/latest/remote_gpio.html) and [PWM](https://gpiozero.readthedocs.io/en/latest/api_output.html#pwmoutputdevice), so maybe this *is* possible, but you'll be on your own with this as GpioZero doesn't have support for NeoPixels yet (see #583 ).

Perhaps something like https://blog.adafruit.com/2022/12/09/control-neopixel-leds-with-raspberry-pi-pico-w-adafruit-io-piday-adafruitio-diyprojectslab-raspberry_pi/ is the kind of thing you're looking for?

> Maybe I make myself illusions but it would be cool to have a "real" editor like "Thonny" on the PC ,instead of "Nano" on the Pi without
having to plug a keyboard mouse and monitor to the Pi (to run Thonny on the Pi)

Have you looked at using [Raspberry Pi Connect](https://www.raspberrypi.com/documentation/services/connect.html) to access your Pi's desktop (including Thonny) from your PC? :slightly_smiling_face:
